Draft HardForks.md (#1401)
IntersectMBO/ouroboros-consensus: docs/website/contents/for-developers/HardForks.md

However, we strongly believe that overall development and assurance would significantly degrade without the benefits of such precise types.

The first era transition (from Byron to Shelley) is a notable example.
It was the first era transition we experienced, and no subsequent transition changes nearly so much.
The two code bases are drastically different, so it's hard to argue this should have been an intra-era hard fork.
In particular, this is the only era transition that changed the duration of slots and the number of slots in an epoch.
In the current codebase, any hard fork that changes those per-era constants must be an era transition, since the Consensus infrastructure was explicitly designed with that restriction as a simplifying assumption.
A different design might relax that constraint, but it has not been restrictive so far: a second change to these constants has been discussed only rarely and never actually proposed, and deferring them until whatever justifies the next era transition wouldn't obviously be prohibitively disruptive.

### How wide-reaching are eras?

Eras are ultimately an implementation detail of the node, a consequence of architectural trade-offs.

> The node's behaviors currently reflects the eras in only two ways.

Reviewer: I would expect these two ways to be mentioned after this sentence :)

Specifically, the node must be able to forecast the data necessary to validate a useful prefix of some header chain from the intersection of that header chain and the node's current selection, including whether each header arrived too early.

> - The familiar and intuitive notion of era is useful for organization in general, and so the eras' names might appear within parameter names in configuration files, specification documents, announcements, etc.

Reviewer: This bullet seems a bit disconnected from the preceding paragraph.

  (See [this GitHub comment](https://github.com/IntersectMBO/ouroboros-consensus/issues/416#issuecomment-2669347315) for the details of how this lesson was learned.)

- Some codecs reflect Cardano's specific era structure.
  That's not strictly necessary, but a tagged union approach has made it trivial to combine the era-specific codecs into a full Cardano codec.
  Specifically, top-level codecs fundamentally branch on a simple number located near the beginning of the bytestring and dispatch accordingly to the corresponding era's codecs.

  *Remark*.
  These eras tags are not present in any codecs for bytes that determine hashes used on chain.
  For example, suppose the only difference between some transaction's serialization in two different eras is the numeric value of the era tag.
  In that case --- unless the eras use different hashing functions --- both serializations would be assigned the same transaction ID, since the function that assigns transaction IDs is applied to the bytes inside of the envelope that carries the era tag.

## Forecasting, forewarning

Hard forks can affect the validity of block headers, and so must be sufficiently forecastable.

> For this reason, hard forks must not happen less than 36 hr (on mainnet) after the voting deadline, ie hard forks must have at least 36 hrs of forewarning --- just like other protocol parameter updates that affect headers (eg maximum block size).

Reviewer: Maybe clarify that 36 hours correspond to 3k?

> For this reason, hard forks must not happen less than 36 hr (on mainnet) after the voting deadline, ie hard forks must have at least 36 hrs of forewarning --- just like other protocol parameter updates that affect headers (eg maximum block size).

Reviewer: ```suggestion
For this reason, hard forks must not happen until 36 hours (on mainnet) after the voting deadline has passed, ie hard forks must have at least 36 hours of forewarning --- just like other protocol parameter updates that affect headers (eg maximum block size).
```

# Hard Forks in Cardano

## Introduction

This document details how the Cardano node handles a [hard fork](https://en.wikipedia.org/wiki/Fork_(blockchain)#Hard_fork).

> Cardano mainnet is intended to hard fork when and only when the major component of the [`protocol version` protocol parameter](https://github.com/IntersectMBO/cardano-ledger/blob/2ff5fa4e8c6b773748799981ef44ef6a62cd5c92/libs/cardano-ledger-core/src/Cardano/Ledger/Core/PParams.hs#L350) changes.

Author: CIP-1694 says "While they need to be treated specially by SPOs, hard forks are not differentiated from other protocol parameter changes."

That's the only difference between `HardForkInitiation` and `UpdateParameters`, right? That `HardForkInitiation` explicitly signals that SPOs need to be paying attention?

> `HardForkInitiation` is also used for updating minor protocol version, albeit it would not be considered a hard fork.

😬

> Cardano mainnet is intended to hard fork when and only when the major component of the [`protocol version` protocol parameter](https://github.com/IntersectMBO/cardano-ledger/blob/2ff5fa4e8c6b773748799981ef44ef6a62cd5c92/libs/cardano-ledger-core/src/Cardano/Ledger/Core/PParams.hs#L350) changes.

Author: Nevermind... I'm already seeing some additional special cases for `HardForkInitiation` within CIP-1694.

> Cardano mainnet is intended to hard fork when and only when the major component of the [`protocol version` protocol parameter](https://github.com/IntersectMBO/cardano-ledger/blob/2ff5fa4e8c6b773748799981ef44ef6a62cd5c92/libs/cardano-ledger-core/src/Cardano/Ledger/Core/PParams.hs#L350) changes.

Reviewer 2: Yeah, SPOs for instance do not even vote on some of the protocol parameter updates and for those that they do vote on, there is a mechanism to override the default vote. Which is not the case with HardForkInitiation. Moreover they have different settings on the required thresholds. So, yeah totally different governance actions

- The requirement of on-chain governance for each hard fork means that --- as of the arrival of decentralized governance to the Cardano mainnet chain --- it's the community that decides when to transition to the next iteration of the protocol rules and/or ledger rules, ie when to abandon any nodes that have not yet updated their software in preparation for the hard fork.

There are other Cardano networks besides mainnet, notably the testnets.
By their essential nature, though, testnets should also use mainnet's mapping between each major protocol version number and the corresponding pair of protocol rules and ledger rules.
On the other hand, these other networks must be able to hard fork at different times than mainnet does --- the testnets should have hard forked successfully before mainnet hard forks!
Since the testnet governance and mainnet governance are independent, the determination of the hard fork timing via on-chain governance immediately enables it to vary on different networks' chains.

## Eras

> A hard fork is the result of the nodes' developers introducing backwards-incompatible changes [^change-decisions] and the community ultimately choosing to adopt them by increasing the major protocol version parameter via on-chain governance [^frame-rule].

Reviewer 2: It is important to note that hard forks happen atomically across the world. That is the whole reason why we need them. SPOs voting for them acts only acts as an indicator that majority of nodes have upgraded their software and are in agreement to do the protocol version bump, while the actual act of bumping the protocol version, i.e. the act of hard forking, happens at one particular instant for everyone.
```suggestion
A hard fork is the result of the nodes' developers introducing backwards-incompatible changes [^change-decisions] and the community ultimately voting via on-chain governance [^frame-rule] to indicate the majority readiness to adopt those changes at the same time by atomically increasing the major protocol version parameter for everyone at once.
```

When those developers design and implement such changes, they determine whether the changes are easy to maintain as variations within the existing code.
If so, then the hard fork is ultimately toggling some conditionals within the code.

> If the changes are not so simple, the Cardano developers instead introduce a new _era_ within the code [^roadmap-disambiguation].

Reviewer 2: It is true, that most of the time it is complexity of the feature that drives the new era, but it is not a requirement. It is not necessarily the simplicity that drives this decision of including a change in an intra-era hard fork. The change could be as simple as prevent duplicates in a Set when it is being deserialized. But because of current limitations of our implementation we do require such change be introduced on a new era. This is a concrete example that we are planning for the next era that we can't introduce with an intra-era hard fork.
Another concrete example was Allegra era. The only change it had was introduction of a Timelock (two extra cases to a native script and a validity interval). It was a fairly simple change, but it required a new era.

The Ledger and Consensus infrastructure code is defined so that each era can flexibly reuse whichever of the older eras' features it does not supersede.
Several eras coexist sequentially on the Cardano chain; a hard fork that takes one step within that sequence is called an _era transition_.

Every hard fork is either an era transition or an _intra-era hard fork_ [^accidentals].
Intra-era hard forks only introduce changes with a small effect on the code, eg a few easy-to-maintain conditionals.
Some have been bugfixes that were small by chance, and others have been changes that were only small because they had been anticipated and intentionally deferred when the era was originally designed.

Eras are a straight-forward mechanism that enables node developers to separate the concerns arising from features that are added to Cardano over time.
The separation amongst eras inherently simplifies the development of the new features, and it also helps ensure that the Cardano node can sync the entire historical chain [^mithril] --- as well as refuse invalid alternative histories.

> ### Which changes require a new era?

Reviewer 2: Another reason for a new era is that sometimes it happens that new features that are being introduced require an initial value, which needs to be populated upon a hard fork. For example initial Constitution and the guardrails script in Conway. Such initial values are supplied through a Genesis file.

> It's ultimately the judgment of the node developers whether some changes should be achieved with an era transition rather than an intra-era hard fork.

Author: @lehins I'd appreciate your feedback/contributions on the entire `Era` section, but especially within this `Which changes require a new era?` subsection.

Author: I had a sentence about "structural changes" specifically, but I seemed to have lost it 🤷

Author: > Moreover, in my personal opinion, this limitation of requiring a hard fork for serialization changes is also a feature. Because serialization changes have a very serious knock-on effect. All of the downstream tooling and hardware wallets are affected by serialization changes. For example we had to freeze all CDDL changes about 6 months before the Conway release, just so hardware wallets were ready for the hard fork.

Did you mean to write "requiring an era for serialization changes"? Otherwise, I'm not seeing how this explicitly relates to the question of whether some hard fork needs to be an era transition rather than "just" an intra-era hard fork.

> changes to a transaction format require changes to plutus context, thus requiring a new plutus ledger version.

I didn't follow these statements about Plutus. But I'm pretty naive about it. Are you ultimately referring to these data structures?

- https://plutus.cardano.intersectmbo.org/haddock/latest/plutus-ledger-api/src/PlutusLedgerApi.V1.Contexts.html#TxInfo
- https://plutus.cardano.intersectmbo.org/haddock/latest/plutus-ledger-api/src/PlutusLedgerApi.V2.Contexts.html#TxInfo
- https://plutus.cardano.intersectmbo.org/haddock/latest/plutus-ledger-api/src/PlutusLedgerApi.V3.Contexts.html#TxInfo

I see how some eras have affected them (eg Conway obviously shows up in V3), but I'm not seeing how the CDDL details of a tx would affect these types, eg.

I chased from `PlutusWithContext` https://github.com/IntersectMBO/cardano-ledger/blob/9f98a9e15d8220ccc4940935efdce8dc69a63c22/libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/Evaluate.hs#L94-L117 and those `TxInfo`s seemed like the only ultimate data that was relevant. Maybe I missed a problematic hash or something?

Maybe the `CostModel? Its Haddock does explicitly mention hard fork boundaries.

Reviewer 2: > Did you mean to write "requiring an era for serialization changes"?

Yes. Sorry, that is exactly what I meant. 

> Are you ultimately referring to these data structures?

Correct

> I'm not seeing how the CDDL details of a tx would affect these types, eg.

It is an indirect relationship that stems from change to Tx, rather than from CDDL. Change to Tx requires CDDL changes and it requires similar change to Plutus Context, thus new Plutus version. There can also be direct relationship, for example we are planning on making inputs an ordered Set, which will affect both CDDL and Plutus context in interesting ways.

Cost Model nowadays (starting with Conway) can be updated at any point in time, so that is not really related to any future hard fork anymore
See this file, this is where era + plutus context mapping happens for Conway (there is similar functionality for previous eras): https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs

The node's behaviors currently reflects the eras in only two ways.

Specifically, the node must be able to forecast the data necessary to validate a useful prefix of some header chain from the intersection of that header chain and the node's current selection, including whether each header arrived too early.

- The familiar and intuitive notion of era is useful for organization in general, and so the eras' names might appear within parameter names in configuration files, specification documents, announcements, etc.

  However, the era transitions themselves should generally be considered as "just a big hard fork".
  The foundational data is instead the protocol version, especially because not all hard forks are era transitions.
  (See [this GitHub comment](https://github.com/IntersectMBO/ouroboros-consensus/issues/416#issuecomment-2669347315) for the details of how this lesson was learned.)

- Some codecs reflect Cardano's specific era structure.
  That's  not strictly necessary, but a tagged union approach has made it trivial to combine the era-specific codecs into a full Cardano codec.

>   Specifically, top-level codecs fundamentally branch on a simple number located near the beginning of the bytestring and dispatch accordingly to the corresponding era's codecs.

Author: > If you are talking about deserializers here then there is no simple number that is added today to the beginning of any block or a transaction that does this sort of branching, nor should it.

That number is present when the block is on the network or stored on disk, etc. We call it "the HFC envelope". It is, by design, transparent to the ledger (the ledger rules generally don't need to consider other eras, except for era-translation functions eg), which is hopefully the only reason this text sticks out to you?

> Protocol version in the ledger state decides the era 

In today's node, the HFC ultimately determines what era some received block/header/transaction _claims_ to be in (according to its HFC envelope). The ledger state determines whether that claim is correct :)

>   Specifically, top-level codecs fundamentally branch on a simple number located near the beginning of the bytestring and dispatch accordingly to the corresponding era's codecs.

Author: I should elaborate on that last quip.

- The HFC tracks which era each ledger state is in. When Consensus ticks a ledger state across an epoch boundary that increments the major protocol version, then the ticked ledger state is now in the next era. (Modulo that block counting wart we've discussed before.)

- The HFC errors out when applying a block that claims to be in some era to a ledger state that is in an older era; the ledger rules aren't even invoked in this case: https://github.com/IntersectMBO/ouroboros-consensus/blob/b14da83d982224edc1700b0bed31ebd96c9ceaf8/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs#L77-L82

>   Specifically, top-level codecs fundamentally branch on a simple number located near the beginning of the bytestring and dispatch accordingly to the corresponding era's codecs.

Author: > Protocol version in the ledger state decides the era and deserializers can use this information to do the branching.

That's the hypothetical we had discussed in the Slack thread: that the Consensus layer might be able to provide the protocol version to deserializers. Currently it does not; instead, the HFC envelope tells it which era's parser to use.

>   Specifically, top-level codecs fundamentally branch on a simple number located near the beginning of the bytestring and dispatch accordingly to the corresponding era's codecs.

Author: > while the actual act of bumping the protocol version, i.e. the act of hard forking, happens at one particular instant for everyone

^^^ that quote is from a different comment thread

> Protocol version in the ledger state decides the era and deserializers can use this information to do the branching.

I fully acknowledge that I can't read your mind, but these two quotes together seem to indicate that you're envisioning that the node parses incoming blocks according to the protocol version of the node's most recent ledger state.

We can't do that, since arriving blocks might not extend the node's currently selection chain. In particular, if an arriving block extends an earlier prefix of the node's selection, then the block to be parsed might be in the "previous" era.

> - (TODO Confirm whether these tags dont affect any hashes. If they do, to what extent does it actually matter?)

Reviewer 2: I am not aware of any tags. As such no hashes are affected.

Author: Ha! Proof by "I've never heard of that" :D Excellent.

  Its latest revision can be accessed via the hyperlink in the "Formal Specification" column of the "Shelley" row of the documentation table at the top of the [README](https://github.com/IntersectMBO/cardano-ledger/blob/master/README.md) of the [`cardano-ledger` repository](https://github.com/IntersectMBO/cardano-ledger).

[^change-decisions]: *Reference*.
  Nowadays, these are almost always going to be (possibly indirectly) related to a Cardano Improvement Proposal (aka [CIP](https://cips.cardano.org/)).

[^frame-rule]: *Clarification*.
  There's no other technical reason for the community to increment the major protocol version.

[^roadmap-disambiguation]: *Disambiguation*.
  The [Cardano Roadmap](https://roadmap.cardano.org/en/) also defines some very broad "eras": Byron, Shelley, Goguen, Basho, and Voltaire.
  These are more general than the eras that divide the actual evolution of the protocol rules and ledger rules within the node.
  [CIP-0059](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0059/feature-table.md) maintains a table showing the correspondence; it lists roadmap eras in the "Phase" column.

[^accidentals]: *Clarification*.

>   Accidental hard forks would necessarily be intra-era hard forks.

Reviewer 2: What is accidental hard fork? Are you talking about an uncontrolled hard fork due to divergence in validity of a block?
If that is the case then it can't be considered as an intra-era hard fork either, IMHO, because no protocol version update would happen. 

However, if you are talking about _accidental_ fork as it is described on wikipedia, then it is not even classified as a *hard* fork, it is part of the normal process of chain selection, which has a potential to span over era boundaries, as far as I understand it.

Author: Hmm. Yeah, I was taking the "intra-era" part too literally. It's a good point that what we call an _intra-era hard fork_ necessarily involves a major protocol version bump, and that wouldn't be accidental.
